docs: add a developer guideline (#1410)
tendermint/starport: docs/developer-guide/hello-world.md

---
order: 5
description: Say "Hello World"
---

### Say "Hello, Starport"

> To get your Cosmos SDK blockchain to say "Hello", you will need to modify a protocol buffer file, create a keeper query function that returns data, and register a query function. Protocol buffer files contain proto `rpc`s that define Cosmos SDK queries and message handlers and proto `message`s that define Cosmos SDK types. `rpc`s are also responsible for exposing an HTTP API. [Keeper](https://docs.cosmos.network/v0.42/building-modules/keeper.html) is an abstraction for modifying the state of the blockchain and keeper functions let you query or write to the state. Registering a query needs to happen only once after you add the first query to your chain.

Reviewer: @barriebyron yeah this seems reasonable. If it gets too over whelming - people loose attention.

> In terms of workflow, developers usually work with proto files first to define Cosmos SDK [messages](https://docs.cosmos.network/v0.42/building-modules/msg-services.html), [queries](https://docs.cosmos.network/v0.42/building-modules/query-services.html), message handlers and then implement the logic of these queries and message handlers in keeper functions.

Reviewer: Proto Buffers (.proto) files help in serialization (transfer) of data. The definition of data to be serialized is written in configuration .proto files.

Let's start by creating a `posts` query:

```
starport scaffold query posts --response title,body
```

`query` accepts a name of the query (in our case, `posts`), an optional list of request parameters (in our case, empty), and an optional comma-separated list of response fields with a `--response` flag (in our case, `body,title`).

The `query` command has created and modified several files:

- modified `proto/blog/query.proto`
- created `x/blog/keeper/grpc_query_posts.go`
- modified `x/blog/client/cli/query.go`

> - created `x/blog/client/cli/query_posts.go`

Reviewer: I think we can let it be as it is. It is just an o/p generated when you scaffold list.

<img width="588" alt="Screen Shot 2021-07-26 at 9 07 21 PM" src="https://user-images.githubusercontent.com/44069404/127078929-14ca7f58-bd7d-4fc9-8f2a-62878146103b.png">

> - created `x/blog/client/cli/query_posts.go`

Author: @barriebyron as @alijnmerchant21 mentioned, keeping it together makes more sense :)

Below you can see both request and response types.

```
message QueryPostsRequest {
}

message QueryPostsResponse {
  string title = 1;
  string body = 2;
}
```

`QueryPostsRequest` is empty because requesting all posts doesn't require are parameters. `QueryPostsResponse` contains `title` and `body` that will be returned from the chain.

`x/blog/keeper/grpc_query_posts.go` contains `Posts` keeper function that handles the query and returns data.

```go
func (k Keeper) Posts(c context.Context, req *types.QueryPostsRequest) (*types.QueryPostsResponse, error) {
  if req == nil {
    return nil, status.Error(codes.InvalidArgument, "invalid request")
  }
  ctx := sdk.UnwrapSDKContext(c)
  _ = ctx
  return &types.QueryPostsResponse{}, nil
}
```

`Posts` function makes a basic check on the request and throws an error if it's `nil`, stores context in a `ctx` variable (context contains information about the environment of the request) and returns a response of type `QueryPostsResponse`. Right now the response is empty.

> From the `query.proto` we know that response may contain `title` and `body`, so let's modify the last line of the function to return a "Hello!".

Reviewer 2: we all use "may" when we mean "can" (or sometimes "might") 
see https://developers.google.com/style/word-list#letter-m 

may
In general, reserve for official policy or legal considerations. To convey possibility, use can or might instead. To convey permission, use can instead. See also can, might, and should.

```go
func (k Keeper) Posts(c context.Context, req *types.QueryPostsRequest) (*types.QueryPostsResponse, error) {
  //...
  return &types.QueryPostsResponse{Title: "Hello!", Body: "Starport"}, nil
}
```

> If we start our chain right now and visit the posts endpoint, we would get a "Not Implemented" error. To fix that we need to wire up our API by registering query handlers with gRPC.
> 
> Inside `x/blog/module.go` import `"context"`, search for `RegisterGRPCGatewayRoutes` and register query handlers:

Reviewer 2: ```suggestion
If you start your chain right now and visit the posts endpoint, you see a "Not Implemented" error. 

To populate the API endpoint, you must register the query handlers with gRPC.

Inside the `x/blog/module.go` file, import `"context"`, search for `RegisterGRPCGatewayRoutes`, and register query handlers:
```

Reviewer 2: I like the term "wire up your API" but does this language work for all of our users? maybe use plain language instead?

```go
import (
  //...
  "context"
)

func (AppModuleBasic) RegisterGRPCGatewayRoutes(clientCtx client.Context, mux *runtime.ServeMux) {
  types.RegisterQueryHandlerClient(context.Background(), mux, types.NewQueryClient(clientCtx))
}
```

> Now we're ready to start our blockchain:

Reviewer 2: ```suggestion
Now, you are ready to start your blockchain:
```

```go
starport chain serve
```

> Once the chain has been started, visit [http://localhost:1317/alice/blog/blog/posts](http://localhost:1317/alice/blog/blog/posts) and see our text displayed!

Reviewer 2: ```suggestion
After the chain has been started, visit [http://localhost:1317/alice/blog/blog/posts](http://localhost:1317/alice/blog/blog/posts) to see your text displayed!
```

Reviewer 2: another one of our favorites is using "once" when we mean "after"
https://developers.google.com/style/word-list?hl=en#letter-o

Reviewer: This might be an edge case - but sometimes the `starport chain serve` won't work because the port is already in use. Do you guys think its in the scope of this tutorial to explain how to free up a port or change the toml file?

```go
{
  "title": "Hello!",
  "body": "Starport"
}
```

The `query` command has also scaffolded `x/blog/client/cli/query_posts.go` that implements a CLI equivalent of the posts query and mounted this command `x/blog/client/cli/query_posts.go` . Run the following command and get the same JSON response:

```go
blogd q blog posts

> ```

Reviewer 2: did you want to show a screen shot of the text displayed? provide a summary of the learning points? See https://www.notion.so/allinbits/Tutorial-standards-62b31701f5124b7989f2cdaa01350a08

(we'll have a tuto template soon)
